catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementTouchListener.java

```diff
@@ -23,50 +23,88 @@
 
 package org.catrobat.catroid.visualplacement;
 
-import android.os.SystemClock;
 import android.view.MotionEvent;
 import android.widget.ImageView;
 
+import java.util.ArrayList;
+import java.util.List;
+
 public class VisualPlacementTouchListener {
 
 	private Mode mode;
-	private float previousY;
+	private static final long DELAY_THRESHOLD = 30;
+	private static final float DISTANCE_THRESHOLD = 10.f * 10.f;
 	private float previousX;
-	private long lastTimeMove = 0;
+	private float previousY;
+	private long touchDownTimestamp = 0;
+	private TimeProvider timeProvider;
+	private List<TouchEventData> eventData = new ArrayList<>();
+
+	private final class TouchEventData {
+		private final long timeStamp;
+		private final float xCoordinate;
+		private final float yCoordinate;
+
+		private TouchEventData(long time, float x, float y) {
+			timeStamp = time;
+			xCoordinate = x;
+			yCoordinate = y;
+		}
+	}
 
 	private void setMode(Mode mode) {
 		this.mode = mode;
 	}
 
+	public VisualPlacementTouchListener(TimeProvider timeProvider) {
+		this.timeProvider = timeProvider;
+	}
+
 	public boolean onTouch(ImageView imageView, MotionEvent event, CoordinateInterface coordinateInterface) {
 		if (event.getPointerId(0) == 0) {
 			float currentX = event.getRawX();
 			float currentY = event.getRawY();
+			long currentTimestamp = timeProvider.elapsedRealtime();
 
 			switch (event.getAction()) {
 				case MotionEvent.ACTION_DOWN:
 					setMode(Mode.TAP);
 					previousX = currentX;
 					previousY = currentY;
-					lastTimeMove = SystemClock.elapsedRealtime();
+					touchDownTimestamp = currentTimestamp;
+					eventData.add(new TouchEventData(currentTimestamp, currentX, currentY));
 					break;
 				case MotionEvent.ACTION_CANCEL:
 				case MotionEvent.ACTION_UP:
 					if (mode == Mode.TAP) {
 						imageView.setX(event.getX() - (float) imageView.getWidth() / 2);
 						imageView.setY(event.getY() - (float) imageView.getHeight() / 2);
+					} else {
+						removeOldEvents(currentTimestamp);
+						if (!eventData.isEmpty()) {
+							TouchEventData oldestEntry = eventData.get(0);
+							float dX = previousX - oldestEntry.xCoordinate;
```

Comment:
Thats the point I was talking with you about on friday, between the last movement and the up event is about ~5 ms so previousX is in 99% of the cases the same as currentX.
From the tests I made I can see that  in rare cases I get a deviation on one parameter of about 0.5 at most. Only if I remove the finger really fast with a swipe movement after holding it I get noticable deviations and as we discussed this isn't really a behaviour we want to correct with this anyways.

I think because of the small delay between the last movement and the up it wouldn't really make a noticable difference if we use the previousX or the currentX.

Reply:
I would still use currentX/Y, because it is more likely that these values then will be congruent with the values given by getX/Y in lines 84-85 below. In that way, there can never be any drift. It's on the safe side to use currentX/Y, even if in 99% of the cases it should be the same as previousX/Y. But with currentX/Y, it is 100%, so why not use it?

Reply:
if we use previous it is always congruent with the values given by getX/Y in lines 84-85
We would have to set them before we get to this point in the programm on the current values, I changed it to current but I am not sure if a deviating value for the current Value here wouldn't be the same as an unwanted jitter right when lifting the finger.